### ISSUE TITLE
tf.keras.layers.Bidirectional is not equivalent to tf.nn.bidirectional_dynamic_rnn

### ISSUE THREAD

**Issue by erikchwang** [tensorflow/tensorflow#26974]
In tf.nn.bidirectional_dynamic_rnn, we can set "sequence_length", which contains the actual lengths of the sequences in a batch. But there is no such setting in tf.keras.layers...
So, if each sample in a batch has a different actual length, how can I do bidirectional_dynamic_rnn in tf.keras.layers?

**Comment by foxik** [comment]
Use masking -- either directly though `tf.keras.layers.Masking`, or through 
`mask_zero` parameter of `tf.keras.layers.Embeddings`. As for masking, Keras has a notion of "masked elements", which are ignored by all subsequent layers (and also losses, so you no longer have to pass weights manually generated using `tf.sequence_mask`).

**Comment by chwang85** [comment]
import tensorflow as tf
inputs = tf.ones([2,10,5])
outputs = tf.nn.bidirectional_dynamic_rnn(cell_fw=tf.nn.rnn_cell.LSTMCell(3),cell_bw=tf.nn.rnn_cell.LSTMCell(3),inputs=inputs,sequence_length=[6,10],dtype=tf.float32)

How can I translate the above code into tf.keras.layers style?

**Comment by chwang85** [comment]
It seems that tf.keras.layers.RNN.call() accepts a "mask" argument, how can I use this argument to solve this problem? The document is not clear at all.

**Comment by chwang85** [comment]
tf.keras.layers.Bidirectional(
  tf.keras.layers.LSTM(
    units=5,
    dropout=0.5,
    return_sequences=True
  )
).apply(
  inputs=tf.ones([2,10,5]),
  mask=tf.expand_dims(input=tf.sequence_mask([6, 10]), axis=2),
  training=True
)

**Comment by chwang85** [comment]
> Use masking -- either directly though `tf.keras.layers.Masking`, or through
> `mask_zero` parameter of `tf.keras.layers.Embeddings`. As for masking, Keras has a notion of "masked elements", which are ignored by all subsequent layers (and also losses, so you no longer have to pass weights manually generated using `tf.sequence_mask`).

It seems that tf.keras.layers.Masking will mask the whole time-step? But what if each sample has different actual length?

**Comment by foxik** [comment]
I would convert
```python
outputs = tf.concat(
  values=tf.nn.bidirectional_dynamic_rnn(
    cell_fw=tf.nn.rnn_cell.LSTMCell(5),
    cell_bw=tf.nn.rnn_cell.LSTMCell(5),
    inputs=tf.ones([2,10,5]),
    sequence_length=[6,10],
    dtype=tf.float32
  )[0],
  axis=2
)
```
to
```python
model = tf.keras.Sequential([
    tf.keras.layers.Masking(mask_value=0.),
    tf.keras.layers.Bidirectional(tf.keras.layers.LSTM(5, return_sequences=True), merge_mode="concat"),
])

inputs = tf.ones([2, 10, 5])
mask = tf.expand_dims(tf.sequence_mask([6, 10], dtype=tf.float32), axis=-1)

print(model(inputs * mask))
```
Or you can use the `mask` directly using
```python
model = tf.keras.Sequential([
    tf.keras.layers.Bidirectional(tf.keras.layers.LSTM(5, return_sequences=True), merge_mode="concat"),
])

inputs = tf.ones([2, 10, 5])
mask = tf.expand_dims(tf.sequence_mask([6, 10], dtype=tf.float32), axis=-1)
print(model(inputs, mask=mask))
```

The `expand_dims` can usually be avoided, because inputs are usually two-dimensional, i.e., `[2, 10]`, and you expand them to `[2, 10, 5]` using an embedding layer.

**Comment by chwang85** [comment]
But they are still not equivalent, because tf.keras.layers dose not support variable scope
See #27016

**Comment by foxik** [comment]
Yes, they are not equivalent. But you can achieve masking, which seemed to me to be the main point of your issue. Personally, I like the TF 2.0/Keras way of masking (because they extend automatically to all following layers and also losses and metrics).

**Comment by chwang85** [comment]
I am a heavy user of "graph + session", maybe I should change my mind to fit into the new TF.

**Comment by chwang85** [comment]
About masking, any suggestion on #27010?

**Comment by ymodak** [comment]
Closing this issue since the original issue has been addressed. Feel free to reopen if have any further questions. Thanks!
